README.md
# blue_thermal_printer
### 支持Xprinter

fork from https://github.com/kakzaki/blue_thermal_printer.git

some code from https://github.com/e3m-software/bluetooth_x_print.git

[![pub package](https://img.shields.io/pub/v/blue_thermal_printer.svg)](https://pub.dev/packages/blue_thermal_printer)
[![pub package](https://img.shields.io/twitter/follow/kakzaki_id.svg?colorA=1da1f2&colorB=&label=Follow%20on%20Twitter)](https://twitter.com/kakzaki_id)

A new Flutter plugin for connecting to thermal printer via bluetooth
(Android Only),
this plugin is still under development
Note: Migrate your project to AndroidX


## Platform Support

| Android | iOS | MacOS | Web | Linux | Windows |
| :-----: | :-: | :---: | :-: | :---: | :-----: |
|   ✔️    | ️X  |  ️X   | ️X  |  ️X   |   ️X    |



## Getting Started

1. Depend on it

Add this to your package's pubspec.yaml file:

    dependencies:
      blue_thermal_printer: ^any

2. Install it

You can install packages from the command line:

with Flutter:

    $ flutter packages get

Alternatively, your editor might support flutter packages get. Check the docs for your editor to learn more.


3. Import it

Now in your Dart code, you can use:

     import 'package:blue_thermal_printer/blue_thermal_printer.dart';


EXAMPLE:

For detail demonstration please look at [example](https://github.com/kakzaki/blue_thermal_printer/tree/master/example) folder

Note: make sure to upgrade your project to compileSdkVersion 31 or above

If you like my content, please consider buying me a coffee. Thank you for your support!

<a href="https://www.buymeacoffee.com/QP1rCmf5L" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png" alt="Buy Me A Coffee" style="height: 41px !important;width: 174px !important;box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;-webkit-box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;" ></a>

[![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/Z8Z6656JW)

bagi yang dari Indonesia bisa lewat saweria berikut

https://saweria.co/kakzaki


CHANGELOG.md
## 1.2.4
* 支持xprinter

## 1.2.3
* demonstrate using enum for readability
* drawer pin by erica
* fix bug on ios

## 1.2.2
* upgrading gradle

## 1.2.1
* fix android 12 permission

## 1.2.0
* fix web build issue

## 1.1.9
* fix bluetooth listener not working, add additional bluetooth state info, add get connected device function
* Issue #123 fix - app will now automatically disconnect if device is turned off
  thanks to knight-dev

## 1.1.8
* Fix bug on ios build

## 1.1.7

* dartdoc comments and Dart formatter

## 1.1.6

* Add print3Column method
* Add print4Column method
* Support format string

## 1.1.5

* Migrate to Android embedding v2

## 1.1.4

* Fix Error Null-Safety

## 1.1.3

* Fix Cast Error on getBondedDevices

## 1.1.2

* Migrate to null-safety

## 1.1.1

* Support different charset in print methods, thanks to danilof

## 1.1.0

* Add print image bytes method, thanks to mvanvu

## 1.0.9

* Avoiding activity null pointer, thanks to wmattei

## 1.0.8

* fix crash with "Methods marked with @UiThread must be executed on the main thread, thanks to ricardochen

## 1.0.7

* change documentation

## 1.0.6

* add size to printleftright

## 1.0.5

* fix bug

## 1.0.4

* fix bug

## 1.0.3

* Add print left + right

## 1.0.2

* Add Endline in QRCODE

## 1.0.1

* Add Doc.

## 1.0.0

* initial release.


lib/blue_thermal_printer.dart
import 'dart:async';
import 'dart:typed_data';

import 'package:flutter/services.dart';

class BlueThermalPrinter {
  static const int STATE_OFF = 10;
  static const int STATE_TURNING_ON = 11;
  static const int STATE_ON = 12;
  static const int STATE_TURNING_OFF = 13;
  static const int STATE_BLE_TURNING_ON = 14;
  static const int STATE_BLE_ON = 15;
  static const int STATE_BLE_TURNING_OFF = 16;
  static const int ERROR = -1;
  static const int CONNECTED = 1;
  static const int DISCONNECTED = 0;
  static const int DISCONNECT_REQUESTED = 2;

  static const String namespace = 'blue_thermal_printer';

  static const MethodChannel _channel =
      const MethodChannel('$namespace/methods');

  static const EventChannel _readChannel =
      const EventChannel('$namespace/read');

  static const EventChannel _stateChannel =
      const EventChannel('$namespace/state');

  final StreamController<MethodCall> _methodStreamController =
      new StreamController.broadcast();

  //Stream<MethodCall> get _methodStream => _methodStreamController.stream;

  BlueThermalPrinter._() {
    _channel.setMethodCallHandler((MethodCall call) async {
      _methodStreamController.add(call);
    });
  }

  static BlueThermalPrinter _instance = new BlueThermalPrinter._();

  static BlueThermalPrinter get instance => _instance;

  ///onStateChanged()
  Stream<int?> onStateChanged() async* {
    yield await _channel.invokeMethod('state').then((buffer) => buffer);

    yield* _stateChannel.receiveBroadcastStream().map((buffer) => buffer);
  }

  ///onRead()
  Stream<String> onRead() =>
      _readChannel.receiveBroadcastStream().map((buffer) => buffer.toString());

  Future<bool?> get isAvailable async =>
      await _channel.invokeMethod('isAvailable');

  Future<bool?> get isOn async => await _channel.invokeMethod('isOn');

  Future<bool?> get isConnected async =>
      await _channel.invokeMethod('isConnected');

  Future<bool?> get openSettings async =>
      await _channel.invokeMethod('openSettings');

  ///getBondedDevices()
  Future<List<BluetoothDevice>> getBondedDevices() async {
    final List list = await (_channel.invokeMethod('getBondedDevices'));
    return list.map((map) => BluetoothDevice.fromMap(map)).toList();
  }

  ///isDeviceConnected(BluetoothDevice device)
  Future<bool?> isDeviceConnected(BluetoothDevice device) =>
      _channel.invokeMethod('isDeviceConnected', device.toMap());

  ///connect(BluetoothDevice device)
  Future<dynamic> connect(BluetoothDevice device) =>
      _channel.invokeMethod('connect', device.toMap());

  ///disconnect()
  Future<dynamic> disconnect() => _channel.invokeMethod('disconnect');

  ///write(String message)
  Future<dynamic> write(Map<String, Object> args) =>
      _channel.invokeMethod('write', args);

  ///writeBytes(Uint8List message)
  Future<dynamic> writeBytes(Uint8List message) =>
      _channel.invokeMethod('writeBytes', {'message': message});

  ///printNewLine()
  Future<dynamic> printNewLine() => _channel.invokeMethod('printNewLine');

}

class BluetoothDevice {
  final String? name;
  final String? address;
  final int type = 0;
  bool connected = false;

  BluetoothDevice(this.name, this.address);

  BluetoothDevice.fromMap(Map map)
      : name = map['name'],
        address = map['address'];

  Map<String, dynamic> toMap() => {
        'name': this.name,
        'address': this.address,
        'type': this.type,
        'connected': this.connected,
      };

  operator ==(Object other) {
    return other is BluetoothDevice && other.address == this.address;
  }

  @override
  int get hashCode => address.hashCode;
}


android/src/main/java/id/kakzaki/blue_thermal_printer/PrintQRCode.java
package id.kakzaki.blue_thermal_printer;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Base64;

import net.posprinter.utils.DataForSendToPrinterTSC;
import net.posprinter.utils.BitmapToByteData;

import android.util.Log;

import java.util.List;
import java.util.Map;
import java.util.ArrayList;

public class PrintQRCode {
    private static final String TAG = PrintQRCode.class.getSimpleName();

    public static List<byte[]> mapToLabel(Map<String,Object> config, List<Map<String,Object>> configs) {
        List<byte[]> list = new ArrayList<>();

        int width = (int)(config.get("width") == null ? 50 : config.get("width")); // 单位：mm
        int height = (int)(config.get("height") == null ? 30 : config.get("height")); // 单位：mm
        int gap = (int)(config.get("gap") == null ? 2 : config.get("gap")); // 单位：mm
        int qty = (int)(config.get("qty") == null ? 1 : config.get("qty")); // 单位：mm

        list.add(DataForSendToPrinterTSC.sizeBymm(width, height));
        list.add(DataForSendToPrinterTSC.gapBymm(gap, 0));
        list.add(DataForSendToPrinterTSC.cls());
        list.add(DataForSendToPrinterTSC.direction(1));

        // {type:'text|barcode|qrcode|image', content:'', x:0,y:0}
        for (Map<String,Object> m: configs) {
            String type = (String) m.get("type");
            String content = (String) m.get("content");
            int x = (int)(m.get("x") == null ? 0 : m.get("x")); //dpi: 1mm约为8个点
            int y = (int)(m.get("y") == null ? 0 : m.get("y"));
            String font_type = (String) (m.get("font_type") == null ? "TSS24.BF2" : m.get("font_type"));
            int size = (int)(m.get("size") == null ? 10 : m.get("size"));
            int x_multification = (int)(m.get("x_multification") == null ? 1 : m.get("x_multification"));
            int y_multification = (int)(m.get("y_multification") == null ? 1 : m.get("y_multification"));
            int rotation = 0;// (int)(m.get("rotation") == null ? 1 : m.get("rotation"));

            if ("text".equals(type)) {
                list.add(DataForSendToPrinterTSC.text(x, y, font_type, rotation, x_multification, y_multification, content));
//                Log.d(TAG, x + ", " + y + ", " + font_type + ", " + x_multification + ", " + y_multification);
            } else if("barcode".equals(type)) {
                list.add(DataForSendToPrinterTSC.barCode(x, y, "128", 100, 1, rotation, 2, 2, content));
            } else if("qrcode".equals(type)) {
                list.add(DataForSendToPrinterTSC.qrCode(x, y, "M", size, "A", rotation, "M2", "S3", content));
            } else if("image".equals(type)) {
                byte[] bytes = Base64.decode(content, Base64.DEFAULT);
                Bitmap bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
                list.add(DataForSendToPrinterTSC.bitmap(x, y, 0, bitmap, BitmapToByteData.BmpType.Threshold));
            }
        }

        list.add(DataForSendToPrinterTSC.print(qty));

        return list;
    }

}


example/lib/testprint.dart
import 'dart:typed_data';
import 'package:blue_thermal_printer_example/printerenum.dart';
import 'package:flutter/services.dart';
import 'package:path_provider/path_provider.dart';
import 'package:blue_thermal_printer/blue_thermal_printer.dart';
import 'dart:io';
import 'package:http/http.dart' as http;

class LineText {
  LineText(
      {this.type, //text,barcode,qrcode,image(base64 string)
      this.content,
      this.size = 0,
      this.align = ALIGN_LEFT,
      this.weight = 0, //0,1
      this.width = 0, //0,1
      this.height = 0, //0,1
      this.underline = 0, //0,1
      this.linefeed = 0, //0,1
      this.x = 0,
      this.y = 0,
      this.font_type = "TSS24.BF2",
      this.rotation = 0,
      this.x_multification = 1,
      this.y_multification = 1});

  static const String TYPE_TEXT = 'text';
  static const String TYPE_BARCODE = 'barcode';
  static const String TYPE_QRCODE = 'qrcode';
  static const String TYPE_IMAGE = 'image';
  static const int ALIGN_LEFT = 0;
  static const int ALIGN_CENTER = 1;
  static const int ALIGN_RIGHT = 2;

  final String? type;
  final String? content;
  final int? size;
  final int? align;
  final int? weight;
  final int? width;
  final int? height;
  final int? underline;
  final int? linefeed;
  final int? x;
  final int? y;
  final String? font_type;
  final int? rotation;
  final int? x_multification;
  final int? y_multification;

  factory LineText.fromJson(Map<String, dynamic> json) {
    return LineText(
      type: json['type'] as String?,
      content: json['content'] as String?,
      size: json['size'] as int?,
      align: json['align'] as int?,
      weight: json['weight'] as int?,
      width: json['width'] as int?,
      height: json['height'] as int?,
      underline: json['underline'] as int?,
      linefeed: json['linefeed'] as int?,
      x: json['x'] as int?,
      y: json['y'] as int?,
      x_multification: json['x_multification'] as int?,
      y_multification: json['y_multification'] as int?,
      font_type: json['font_type'] as String?,
      rotation: json['rotation'] as int?,
    );
  }

  Map<String, dynamic> toJson(LineText data) {
    final val = <String, dynamic>{};

    void writeNotNull(String key, dynamic value) {
      if (value != null) {
        val[key] = value;
      }
    }

    writeNotNull('type', data.type);
    writeNotNull('content', data.content);
    writeNotNull('size', data.size);
    writeNotNull('align', data.align);
    writeNotNull('weight', data.weight);
    writeNotNull('width', data.width);
    writeNotNull('height', data.height);
    writeNotNull('underline', data.underline);
    writeNotNull('linefeed', data.linefeed);
    writeNotNull('x', data.x);
    writeNotNull('y', data.y);
    writeNotNull('x_multification', data.x_multification);
    writeNotNull('y_multification', data.y_multification);
    writeNotNull('font_type', data.font_type);
    writeNotNull('rotation', data.font_type);
    return val;
  }
}

///Test printing
class TestPrint {
  BlueThermalPrinter bluetooth = BlueThermalPrinter.instance;

  sample() async {
    //image max 300px X 300px

    ///image from File path
    // String filename = 'yourlogo.png';
    // ByteData bytesData = await rootBundle.load("assets/images/yourlogo.png");
    // String dir = (await getApplicationDocumentsDirectory()).path;
    // File file = await File('$dir/$filename').writeAsBytes(bytesData.buffer
    //     .asUint8List(bytesData.offsetInBytes, bytesData.lengthInBytes));
    //
    // ///image from Asset
    // ByteData bytesAsset = await rootBundle.load("assets/images/yourlogo.png");
    // Uint8List imageBytesFromAsset = bytesAsset.buffer
    //     .asUint8List(bytesAsset.offsetInBytes, bytesAsset.lengthInBytes);
    //
    // ///image from Network
    // var response = await http.get(Uri.parse(
    //     "https://raw.githubusercontent.com/kakzaki/blue_thermal_printer/master/example/assets/images/yourlogo.png"));
    // Uint8List bytesNetwork = response.bodyBytes;
    // Uint8List imageBytesFromNetwork = bytesNetwork.buffer
    //     .asUint8List(bytesNetwork.offsetInBytes, bytesNetwork.lengthInBytes);

    bluetooth.isConnected.then((isConnected) {
      if (isConnected == true) {
        Map<String, dynamic> config = Map();
        config['width'] = 50; // 标签宽度，单位mm
        config['height'] = 30; // 标签高度，单位mm
        config['gap'] = 2; // 标签间隔，单位mm

        // x、y坐标位置，单位dpi，1mm=8dpi
        List<LineText> list = [];
        list.add(LineText(
            type: LineText.TYPE_TEXT, x: 140, y: 10, content: 'ABCDEF'));
        list.add(LineText(
            type: LineText.TYPE_TEXT, x: 100, y: 40, content: '一二三四五'));
        list.add(LineText(
            type: LineText.TYPE_QRCODE,
            x: 120,
            y: 80,
            size: 5,
            content: '2J202034234123123324324'));
        Map<String, Object> args = Map();
        args['config'] = config;
        args['data'] = list.map((e) => LineText().toJson(e)).toList();
        print('args --- $args');
        bluetooth.write(args);
        // bluetooth.printNewLine();

        // bluetooth.printCustom("HEADER", Size.boldMedium.val, Align.center.val);
        // bluetooth.printNewLine();
        // bluetooth.printImage(file.path); //path of your image/logo
        // bluetooth.printNewLine();
        // bluetooth.printImageBytes(imageBytesFromAsset); //image from Asset
        // bluetooth.printNewLine();
        // bluetooth.printImageBytes(imageBytesFromNetwork); //image from Network
        // bluetooth.printNewLine();
        // bluetooth.printLeftRight("LEFT", "RIGHT", Size.medium.val);
        // bluetooth.printLeftRight("LEFT", "RIGHT", Size.bold.val);
        // bluetooth.printLeftRight("LEFT", "RIGHT", Size.bold.val,
        //     format:
        //         "%-15s %15s %n"); //15 is number off character from left or right
        // bluetooth.printNewLine();
        // bluetooth.printLeftRight("LEFT", "RIGHT", Size.boldMedium.val);
        // bluetooth.printLeftRight("LEFT", "RIGHT", Size.boldLarge.val);
        // bluetooth.printLeftRight("LEFT", "RIGHT", Size.extraLarge.val);
        // bluetooth.printNewLine();
        // bluetooth.print3Column("Col1", "Col2", "Col3", Size.bold.val);
        // bluetooth.print3Column("Col1", "Col2", "Col3", Size.bold.val,
        //     format:
        //         "%-10s %10s %10s %n"); //10 is number off character from left center and right
        // bluetooth.printNewLine();
        // bluetooth.print4Column("Col1", "Col2", "Col3", "Col4", Size.bold.val);
        // bluetooth.print4Column("Col1", "Col2", "Col3", "Col4", Size.bold.val,
        //     format: "%-8s %7s %7s %7s %n");
        // bluetooth.printNewLine();
        // bluetooth.printCustom("čĆžŽšŠ-H-ščđ", Size.bold.val, Align.center.val,
        //     charset: "windows-1250");
        // bluetooth.printLeftRight("Številka:", "18000001", Size.bold.val,
        //     charset: "windows-1250");
        // bluetooth.printCustom("Body left", Size.bold.val, Align.left.val);
        // bluetooth.printCustom("Body right", Size.medium.val, Align.right.val);
        // bluetooth.printNewLine();
        // bluetooth.printCustom("Thank You", Size.bold.val, Align.center.val);
        // bluetooth.printNewLine();
        // bluetooth.printQRcode(
        //     "Insert Your Own Text to Generate", 200, 200, Align.center.val);
        // bluetooth.printNewLine();
        // bluetooth.printNewLine();
        // bluetooth
        //     .paperCut(); //some printer not supported (sometime making image not centered)
        //bluetooth.drawerPin2(); // or you can use bluetooth.drawerPin5();
      }
    });
  }

//   sample(String pathImage) async {
//     //SIZE
//     // 0- normal size text
//     // 1- only bold text
//     // 2- bold with medium text
//     // 3- bold with large text
//     //ALIGN
//     // 0- ESC_ALIGN_LEFT
//     // 1- ESC_ALIGN_CENTER
//     // 2- ESC_ALIGN_RIGHT
//
// //     var response = await http.get("IMAGE_URL");
// //     Uint8List bytes = response.bodyBytes;
//     bluetooth.isConnected.then((isConnected) {
//       if (isConnected == true) {
//         bluetooth.printNewLine();
//         bluetooth.printCustom("HEADER", 3, 1);
//         bluetooth.printNewLine();
//         bluetooth.printImage(pathImage); //path of your image/logo
//         bluetooth.printNewLine();
// //      bluetooth.printImageBytes(bytes.buffer.asUint8List(bytes.offsetInBytes, bytes.lengthInBytes));
//         bluetooth.printLeftRight("LEFT", "RIGHT", 0);
//         bluetooth.printLeftRight("LEFT", "RIGHT", 1);
//         bluetooth.printLeftRight("LEFT", "RIGHT", 1, format: "%-15s %15s %n");
//         bluetooth.printNewLine();
//         bluetooth.printLeftRight("LEFT", "RIGHT", 2);
//         bluetooth.printLeftRight("LEFT", "RIGHT", 3);
//         bluetooth.printLeftRight("LEFT", "RIGHT", 4);
//         bluetooth.printNewLine();
//         bluetooth.print3Column("Col1", "Col2", "Col3", 1);
//         bluetooth.print3Column("Col1", "Col2", "Col3", 1,
//             format: "%-10s %10s %10s %n");
//         bluetooth.printNewLine();
//         bluetooth.print4Column("Col1", "Col2", "Col3", "Col4", 1);
//         bluetooth.print4Column("Col1", "Col2", "Col3", "Col4", 1,
//             format: "%-8s %7s %7s %7s %n");
//         bluetooth.printNewLine();
//         String testString = " čĆžŽšŠ-H-ščđ";
//         bluetooth.printCustom(testString, 1, 1, charset: "windows-1250");
//         bluetooth.printLeftRight("Številka:", "18000001", 1,
//             charset: "windows-1250");
//         bluetooth.printCustom("Body left", 1, 0);
//         bluetooth.printCustom("Body right", 0, 2);
//         bluetooth.printNewLine();
//         bluetooth.printCustom("Thank You", 2, 1);
//         bluetooth.printNewLine();
//         bluetooth.printQRcode("Insert Your Own Text to Generate", 200, 200, 1);
//         bluetooth.printNewLine();
//         bluetooth.printNewLine();
//         bluetooth.paperCut();
//       }
//     });
//   }
}
